juju-tutorials/juju-tricks.md

```diff
@@ -0,0 +1,177 @@
+```
+
+To restart machine in particular state (including LXCs) use the following command and adjust the first part, `state=<TARGET_STATE>`, to match state of the machines that you want to restart. The command, as it is, will restart machines in state `down`.
+
+```console
+state=down; juju machines | grep -v 'Machine' | awk -v state=${state} '{if ($2 == state) print $1}' | xargs -P4 -I@ bash -c 'machine=@; svc=${machine//\//-}; svc=${svc%:}; juju ssh ${machine%:} "sudo service jujud-machine-${svc}" restart'
+```
+
+### Restarting all units and machines
+
+To restart all units, regardless of their state, run following command.
+
+```console
+juju status --format oneline | xargs -l4 | column -t | egrep -o '[a-z0-9-]+/[0-9]+' | sort -u | xargs -P4 -I@ bash -c 'unit=@; juju ssh $unit "sudo service jujud-unit-${unit/\//-} restart"
+```
+
+To restart all machines, regardless of their state, run following command.
+
+```console
+juju machines | grep -v 'Machine' | awk '{print $1}' | xargs -P4 -I@ bash -c 'machine=@; svc=${machine//\//-}; svc=${svc%:}; juju ssh ${machine%:} "sudo service jujud-machine-${svc}" restart'
+```
+
+## Tricks on debugging unit relations
+
+Duration: N/A
+
+### Inspecting relation data
+
+To see data of the specific relation on certain unit, run the following command and adjust variables `unit` and `relation`  to match unit and relation that you wish to inspect. This example command will print data from the relation `cni` of the unit `kubernetes-master/0`.
+
+```console
+unit=kubernetes-master/0;relation=cni; juju run --unit $unit "relation-ids $relation | xargs -I_@ sh -c 'relation-list -r _@ | xargs -I_U sh -c \"relation-get -r _@ - _U | sed s,^,_U:, 2>&1\"'"
+```
+
+### Manually setting relation data
+
+To manually set data on some relation, we must first find its relation ID. Lets say we want to change value of the `cni-conf-file` on the `cni` relation between units `kubernetes-worker/0` and `flannel/1`. We run this command to determine relation ID.
+
+```console
+juju run --unit kubernetes-master/0 'relation-ids cni'
+cni:11
+```
+
+Then we can set the new value by running
+
+```console
+juju run --unit flannel/1 "relation-set -r cni:11 cni-conf-file=10-flannel.conflist"
+```
+
+### Triggering relation-changed hook
+
+The easiest way to trigger `relation-changed` hook is to change data in the relation using dummy variable that will be ignored by the other end of the relation. See chapter above, "Manually setting relation data" to see how to run `relation-set`.
+
+[note status="Reactive charms"]
+For the reactive charms, use 'juju run --unit foo/0 charms.reactive' to set/unset the flags and wait for the update-status hook to run
+[/note]
+
+## Debugging
+
+Duration: N/A
+
+## Inspecting state of reactive charms
+
+It's real simple to check states of the reactive charm. In the example below we will inspect states of the unit `flannel/1`
```

Comment:
Could we share here a brief reason why this command is useful? States (or flags) are used by the reactive framework to transition the components of a unit (relations, configurations, ...) from a not-ready to a ready status. The command will help troubleshoot a possible bug in the charm.